wgpu-core/src/command/mod.rs

```diff
@@ -78,21 +78,161 @@ pub(crate) enum CommandEncoderStatus {
+        match *self {
+            Self::Recording(_) => {
+                change_state_with_inner!(self, Self::Recording, Self::Locked);
+                Ok(())
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Locked(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Locked)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    /// Unlocks the [`CommandBuffer`] and puts it back into the [`Self::Recording`] state.
+    ///
+    /// This function is the counterpart to [`Self::lock_encoder`].
+    /// It is only valid to call this function if the encoder is in the [`Self::Locked`] state.
+    fn unlock_encoder(&mut self) -> Result<EncoderGuard<'_>, CommandEncoderError> {
+        match *self {
+            Self::Locked(_) => {
+                change_state_with_inner!(self, Self::Locked, Self::Recording);
+                Ok(EncoderGuard {
+                    inner: self,
+                    succeeded: false,
+                })
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Recording(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Invalid)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    fn finish(&mut self, device: &Device) -> Result<(), CommandEncoderError> {
+        match self {
+            Self::Recording(inner) => {
+                if let Err(e) = inner.encoder.close(device) {
+                    let _ = mem::replace(self, Self::Error);
+                    Err(e.into())
+                } else {
+                    change_state_with_inner!(self, Self::Recording, Self::Finished);
+                    // Note: if we want to stop tracking the swapchain texture view,
+                    // this is the place to do it.
+                    Ok(())
+                }
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Locked(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Locked)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+}
+
+/// A guard on which [`EncoderGuard::succeeded`] should be called if
+/// no errors have been raised during encoding, otherwise the encoder will be put in the
+/// [`CommandEncoderStatus::Error`] on [`EncoderGuard::drop`].
+pub(crate) struct EncoderGuard<'a> {
+    inner: &'a mut CommandEncoderStatus,
+    succeeded: bool,
+}
+
+impl<'a> EncoderGuard<'a> {
+    pub(crate) fn succeeded(&mut self) {
+        self.succeeded = true;
+    }
+}
+
+impl<'a> Drop for EncoderGuard<'a> {
+    fn drop(&mut self) {
+        if !self.succeeded {
+            let _ = mem::replace(&mut *self.inner, CommandEncoderStatus::Error);
```

Comment:
suggestion: I think it's easier to understand an assignment:

```suggestion
            *self.inner = CommandEncoderStatus::Error;
```

...unless you intended to ensure that the `Error` state was assigned before a potential panic on `drop`. This would be clever, but should also get a comment to explain the intent:

```suggestion
            // NOTE: We `mem::replace` here to ensure that the `Error` state is set before we
            // attempt to `drop` the old state.
            let _ = mem::replace(&mut *self.inner, CommandEncoderStatus::Error);
```

Pushed the latter suggestion as a `fixup!` (45118766), assuming that this was intended.

Reply:
I can't figure out what harm would arise if we dropped the old state first. For drop order to matter, there's got to be some shared mutable state somewhere, and I don't see it.

Reply:
Doing it via an assignment should work and would be easier to understand. This got copied from the other places where we use `mem::replace` to get to the old `inner` state.